vagrant: Add openSUSE support

vadorovsky opened the pull request:
We already have a Vagrant box for openSUSE and this change brings
the support of them in all Vagrantfiles. To sum it up, we will
support Ubuntu and openSUSE as distributions that developer
can choose for his dev environment and we may use both of them
in CI in the future.

The choice of distribution is determined by the DISTRIBUTION
environment variable. If it's not defined, then Ubuntu will be
chosen. So, for example, if someone wants to use Ubuntu, then our
script for running Vagrant can be used as usually:

```
./contrib/vagrant/start.sh
```

Or we can explicitly tell the script to use Ubuntu:

```
export DISTRIBUTION=ubuntu
./contrib/vagrant/start.sh
```

If someone wants to use openSUSE, then the following commands
need to be used:

```
export DISTRIBUTION=opensuse
./contrib/vagrant/start.sh
```

In order to bring the support of openSUSE, some other changes were
needed.

First of them was changing the directory for systemd units from
/lib/systemd/system to /etc/systemd/system. That's because Ubuntu
and openSUSE use different directories for installing systemd units
from packages. Ubuntu - /lib/systemd/system, openSUSE (and other
RPM-based distributions) - /usr/lib/systemd/system. On the other
hand, /etc/systemd/system directory is expected to be used by
administrators to put hand-written units in and it doesn't differ
in distributions. That's why using /etc/systemd/system is more
appropriate for us.

The other change is related to network interface names. Ubuntu
uses NetworkManager[1] which uses names like i.e. enp0s8.
openSUSE uses wicked[2] which uses names like i.e. eth1.

Usage of go binary for fetching dependencies (go-bindata, gops)
needed to change as well - we're using it from $PATH instead of
hardcoding the absolute path. In Ubuntu we fetch go binary and
save it in /usr/local/go/bin/go, in openSUSE we install go from
packages.

On Ubuntu, test/Vagrantfile needs to modify the mesg command in
/root/.profile, but openSUSE doesn't contain that file at all,
so it's just created with the appropriate content for both
distros.

The last needed change was using useradd instead of adduser.
useradd is the more commonly used project and adduser isn't
packaged in openSUSE (and in the most RPM-based distributions
as well).

[1] https://wiki.gnome.org/Projects/NetworkManager
[2] https://github.com/openSUSE/wicked

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/2712)
<!-- Reviewable:end -->

vadorovsky committed:
vagrant: Add openSUSE support

We already have a Vagrant box for openSUSE and this change brings
the support of them in all Vagrantfiles. To sum it up, we will
support Ubuntu and openSUSE as distributions that developer
can choose for his dev environment and we may use both of them
in CI in the future.

The choice of distribution is determined by the DISTRIBUTION
environment variable. If it's not defined, then Ubuntu will be
chosen. So, for example, if someone wants to use Ubuntu, then our
script for running Vagrant can be used as usually:

	./contrib/vagrant/start.sh

Or we can explicitly tell the script to use Ubuntu:

	export DISTRIBUTION=ubuntu
	./contrib/vagrant/start.sh

If someone wants to use openSUSE, then the following commands
need to be used:

	export DISTRIBUTION=opensuse
	./contrib/vagrant/start.sh

In order to bring the support of openSUSE, some other changes were
needed.

First of them was changing the directory for systemd units from
/lib/systemd/system to /etc/systemd/system. That's because Ubuntu
and openSUSE use different directories for installing systemd units
from packages. Ubuntu - /lib/systemd/system, openSUSE (and other
RPM-based distributions) - /usr/lib/systemd/system. On the other
hand, /etc/systemd/system directory is expected to be used by
administrators to put hand-written units in and it doesn't differ
in distributions. That's why using /etc/systemd/system is more
appropriate for us.

The other change is related to network interface names. Ubuntu
uses NetworkManager[1] which uses names like i.e. enp0s8.
openSUSE uses wicked[2] which uses names like i.e. eth1.

Usage of go binary for fetching dependencies (go-bindata, gops)
needed to change as well - we're using it from $PATH instead of
hardcoding the absolute path. In Ubuntu we fetch go binary and
save it in /usr/local/go/bin/go, in openSUSE we install go from
packages.

On Ubuntu, test/Vagrantfile needs to modify the mesg command in
/root/.profile, but openSUSE doesn't contain that file at all,
so it's just created with the appropriate content for both
distros.

The last needed change was using useradd instead of adduser.
useradd is the more commonly used project and adduser isn't
packaged in openSUSE (and in the most RPM-based distributions
as well).

[1] https://wiki.gnome.org/Projects/NetworkManager
[2] https://github.com/openSUSE/wicked

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

eloycoto commented on test/Vagrantfile:
```diff
@@ -6,8 +6,14 @@ $JOB_NAME = ENV['JOB_BASE_NAME'] || "LOCAL"
 
-$SERVER_BOX= "cilium/ginkgo"
-$SERVER_VERSION="0.0.4"
+$distribution = ENV['DISTRIBUTION']
```
Hi, 

Let me suggest a new way in the test/Vagrantfile: 

I would do something like: 

```
$DISTRIBUTION = ENV['DISTRIBUTION'] || ubuntu
... 
server.vm.box = cilium/$DISTRIBUTION
server.vm.box_version = $SERVER_VERSION
```

The tty line is needed.

On the other hand, both version will be the same, build on packer at the same
time, so will be the same, and are super important for us 

In Jenkins we can call DISTRIBUTION=opensuse and will work as normal. ;-)

vadorovsky replied:
>On the other hand, both version will be the same, build on packer at the same
time, so will be the same, and are super important for us

Sorry, but I don't understand this sentence. What version you're referring to? `SERVER_VERSION`? They seem to be different for Ubuntu and openSUSE:

https://app.vagrantup.com/cilium/boxes/ubuntu (the newest is v0)
https://app.vagrantup.com/cilium/boxes/opensuse (the newest is v1, the oldest is v0.0.1)

Can't we just not define the box version and rely on the newest boxes which are available?

If for some reason (which I'm possibly not aware about) we need to define the server version, then we probably need to define the version for each distro - and keep it in some dictionary, hash table or any kind of data struct we can simply keep in the Vagrantfile.

vadorovsky replied:
Also, your proposed code with `$DISTRIBUTION = ENV['DISTRIBUTION'] || ubuntu` looks simpler, but it will not give an explicit error if user will make a mistake in the distro name.

vadorovsky replied:
Also, the tty line does **not** work on any distro. It seems that it worked for the older Ubuntu releases, but not for 16.10. The `/root/.profile` file just doesn't exist.

eloycoto replied:
Hi, 

The version of the boxes will be fixed today. Both servers will have the same version.  PR 8 in packer-ci-build ;-) 

Related with the version, no we don't want to use the latest, the idea is to have a specific version and be sure that we always use this. In case of a change in a base box, maybe we need to do some work in cilium, so we don't use that until a new PR that updates the version. 

Around the DISTRIBUTION,  if is not correct, the vagrant will give an error that the box does not exit. In the other hand, that options should be documented in the `docs/contributing.rst`. The idea is to have the vagrantfile as clean as possible.

aanm commented on Vagrantfile:
```diff
@@ -134,21 +109,12 @@ Vagrant.configure(2) do |config|
 
-        config.vm.box = "cilium/ubuntu-16.10"
-	config.vm.box_version = "2.7"
+        config.vm.box = $box
```
please specify the box version as well

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -195,9 +194,6 @@ function write_install_nsenter(){
 
 if [ -n "\${INSTALL}" ]; then
     #Install nsenter
-    sudo apt-get -y install bison libncurses5-dev libslang2-dev gettext \
```
You can remove this whole thing from line 196 to line 211 since it is already installed in the new ubuntu box version 8

vadorovsky committed:
Documentation: Update dev guide after recent Vagrant changes

Recently, we changed the following things in our Vagrant
environment:

- we dropped support for libvirt provider
- we added support for openSUSE in form of Vagrant box
- we upgraded Ubuntu to 17.10
- every supported distribution uses systemd

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

eloycoto commented on test/provision/k8s_install.sh:
```diff
@@ -20,7 +20,11 @@ source ${PROVISIONSRC}/helpers.bash
 
 if [[ -f  "/etc/provision_finished" ]]; then
-    sudo dpkg -l | grep kubelet
+    if grep Ubuntu /etc/lsb-release; then
```
I think that this code can be deleted no?

vadorovsky replied:
I don't know. It can be? I didn't introduce anything else there.

eloycoto commented on test/provision/k8s_install.sh:
```diff
@@ -62,10 +75,16 @@ KUBEADM_SLAVE_OPTIONS=""
 case $K8S_VERSION in
-    "1.7"|"1.8")
+    "1.7")
+        K8S_FULL_VERSION="1.7.11"
```
Why is this needed? I would like to have the behaviour that we had in the past.

vadorovsky replied:
Unfortunately, you cannot tell `zypper` to install just `1.7` version. You need to specify the whole version that the concrete RPM package has. That's sad, but I cannot do anything about that.

vadorovsky replied:
It means, that `zypper in kubelet=1.7.11` works and `zypper in kubelet=1.7` doesn't. Same with wildcards.

ianvernon commented on test/provision/k8s_install.sh:
```diff
@@ -20,7 +20,11 @@ source ${PROVISIONSRC}/helpers.bash
+    if grep Ubuntu /etc/lsb-release; then
+        sudo dpkg -l | grep kubelet
+    else
+        sudo rpm -q kubele
```
kubelet?
